lib/spack/spack/cmd/__init__.py

```diff
@@ -85,10 +91,24 @@ def get_module(cmd_name):
             (contains ``-``, not ``_``).
     """
     pname = python_name(cmd_name)
-    module_name = "%s.%s" % (__name__, pname)
-    module = __import__(module_name,
-                        fromlist=[pname, SETUP_PARSER, DESCRIPTION],
-                        level=0)
+
+    try:
+        # Try to import the command from the built-in directory
+        module_name = "%s.%s" % (__name__, pname)
+        module = __import__(module_name,
+                            fromlist=[pname, SETUP_PARSER, DESCRIPTION],
+                            level=0)
+        tty.debug('Imported {0} from built-in commands'.format(pname))
+    except ImportError:
+        # If built-in failed the import search the extension
+        # directories in order
```

Comment:
> Does "in order" matter? In one sense it does: any extension directory can support multiple commands, and so you could have conflicting implementations of a command

"In order" matters for that reason. If there are multiple commands with the same name you'll currently get the first one that is matched. Those kind of conflicts are probably a niche use-case right now, but this seemed a simple and sensible solution at the time of writing the PR. Templates should work the same, even though again I don't think anybody came up with a use case that requires overriding a template completely.

Reply:
> I acknowledge it may be convenient to implement several commands in one extension directory (in particular if they have related functionality that you want to test); so in general I'm OK with having multiple commands in one extension directory.

This was my intent when writing this. An extension is meant to provide functionality for a specific purpose and it can be that more than one command is needed for that.

Reply:
> Also, IMO the next several lines should be placed into a function inside the extensions module.

:+1: